Mission 3 ~ 5 해결

Min-Su opened the pull request:
#4 미션3 {name} 값만 반환하는 기능

Min-Su committed:
WIPG

monorisk commented on src/main/java/com/minsuhas/springboot/controller/mission3Controller.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class mission3Controller {
```
클래스의 이름은 대문자로 시작합니다

monorisk replied:
https://blog.managr.us/entry/%EB%AA%85%EB%AA%85-%EA%B7%9C%EC%B9%99%EB%B3%80%EC%88%98%EB%AA%85-%EA%B7%9C%EC%95%BD-Naming-Rule-Naming-Convention-Java-%EB%AA%85%EB%AA%85%EA%B7%9C%EC%B9%99-HTML-%EB%AA%85%EB%AA%85%EA%B7%9C%EC%B9%99

Min-Su replied:
표기법은 J2EE 표기법으로 숙달 하도록 하겠습니다.

monorisk commented on src/main/java/com/minsuhas/springboot/controller/mission3Controller.java:
```diff
@@ -0,0 +1,15 @@
+@RestController
+public class mission3Controller {
+
+    @RequestMapping(value = "/hello/{name}", method = {RequestMethod.GET, RequestMethod.POST})
```
전부터 궁금 했는데, 왜 method를 따로 지정 하고 있나요?

따로 지정하는 이유가 있다면 왜 GET, POST만 지정 하나요?

Min-Su replied:
해당 부분은 구글링을 통해 참고한 블로그에서 사용한 방법입니다. 다른 사용방법에는 어떤게 있나요?

monorisk replied:
https://docs.spring.io/spring/docs/5.1.5.RELEASE/spring-framework-reference/web.html#mvc-ann-methods

monorisk commented on src/main/java/com/minsuhas/springboot/controller/mission3Controller.java:
```diff
@@ -0,0 +1,15 @@
+public class mission3Controller {
+
+    @RequestMapping(value = "/hello/{name}", method = {RequestMethod.GET, RequestMethod.POST})
+    public String returnString(@PathVariable String name) {
```
메서드의 역할을 좀 더 명확하게 표현 하자면 returnName 처럼 도메인을 잘 설명하는 이름이 나을 것 같습니다.

monorisk replied:
~~도메인~~ -> 도메인 내에서 비지니스

Min-Su committed:
- 클래스 컨밴션 첫 글자 대문자로 치환
 - Http Method 제한 하는 로직 제거
 - 좀 더 명확한 역할을 표현한 메서드 명으로 변경

monorisk commented on src/test/java/com/minsuhas/springboot/controller/Mission3ControllerTest.java:
```diff
@@ -12,14 +12,14 @@
 
     @Test
-    public void returnString() throws Exception {
+    public void returnName() throws Exception {
```
테스트 메서드 컨밴션에 어긋납니다.

Min-Su committed:
- 클래스 컨밴션 첫 글자 대문자로 치환
 - Http Method 제한 하는 로직 제거
 - 좀 더 명확한 역할을 표현한 메서드 명으로 변경
 + - 테스트 메서드명 변경

Min-Su committed:
#mission 3
 - 클래스 컨밴션 첫 글자 대문자로 치환
 - Http Method 제한 하는 로직 제거
 - 좀 더 명확한 역할을 표현한 메서드 명으로 변경
 - 테스트 메서드명 변경

#mission 4
 - mission 4 Image Resource 반환 기능 개발

Min-Su committed:
#mission 5
 - Json 데이터 저장 클래스 User 생성
 - User 클래스 Path로 받은 name set 후 반환하는 기능 개발
 - 해당 Test 코드 작성

monorisk commented on pom.xml:
```diff
@@ -32,7 +32,11 @@
-
+		<dependency>
+			<groupId>com.fasterxml.jackson.dataformat</groupId>
+			<artifactId>jackson-dataformat-xml</artifactId>
```
jackson은 Spring Boot에 묵시적으로 Bean으로 선언 되어 있어 추가적인 의존성 추가가 불필요합니다.

monorisk replied:
필요하다면 Bean 내부에서 다음과 같은 코드로 주입 받을 수 있습니다.
(Thread Safe한 객체이며, 내부 설정은 모든 Thread가 공유 합니다.)

```java
@Autowired
ObjectMapper objectMapper;
```

monorisk commented on src/main/java/com/minsuhas/springboot/controller/Mission3Controller.java:
```diff
@@ -0,0 +1,14 @@
+@RestController
+public class Mission3Controller {
+
+    @RequestMapping(value = "/hello/{name}")
```
이렇게 `produces` 를 지정하지 않으면 `Content-Type`이 `text/html`로 반환 됩니다.

하지만 이 데이터를 보면 알듯이, 이것을 HTML 이라고 부르기에는 문제가 있죠.

따라서 `text/plain`으로 Content-Type을 반환 하는 것을 추천 합니다.

monorisk commented on src/main/java/com/minsuhas/springboot/controller/Mission4Controller.java:
```diff
@@ -0,0 +1,21 @@
+@RestController
+public class Mission4Controller {
+
+    @GetMapping(value = "/resource/hello/image", produces = MediaType.IMAGE_JPEG_VALUE)
```
`@GetMapping`은 HTTP Method들 중, `GET` 방식의 요청만 허용 하는 Annotation 입니다.

모든 HTTP Method를 허용 하고 싶다면, 그냥 `@RequestMapping` 을 사용하면 됩니다.
(GET만 주는 것을 의도 했다면 매우 정확한 구현입니다.)

monorisk commented on src/main/java/com/minsuhas/springboot/controller/Mission4Controller.java:
```diff
@@ -0,0 +1,21 @@
+    @GetMapping(value = "/resource/hello/image", produces = MediaType.IMAGE_JPEG_VALUE)
+    public byte[] returnImage() throws IOException {
+        InputStream in = getClass()
+                .getResourceAsStream("/static/cat_1.jpg");
```
### 추가적인 고민
* 이렇게 고정된 이미지만을 반환하는 경우, 어떤 식으로 개선 하면 File I/O 비용을 줄일 수 있을 까요?
   * InputStream을 통해 byte[]를 얻어낼 때마다 File I/O가 수행 됩니다.

Min-Su replied:
1 byte 단위로 File I/O가 수행되는 InputStream이 아닌 Buffer에 담아 File I/O를 수행하는 BufferedInputStream 사용합니다.

monorisk commented on src/main/java/com/minsuhas/springboot/controller/Mission5Controller.java:
```diff
@@ -0,0 +1,17 @@
+public class Mission5Controller {
+
+    @RequestMapping(value = "/resource/hello/{name}/json")
+    public User returnJson(@RequestBody User user, @PathVariable String name) {
```
어제 같이 해보았듯이, `@RequestBody`는 뜻 그대로 HTTP Request의 Body 영역 안에 있는 JSON을 User 클래스의 인스턴스로 변환 해주는 역할을 합니다.

우리가 의도한 내용은 user객체를 전달 받아 사용하는게 아니라, `PathVariable`에서 이름 값을 알아낸 뒤, User객체에 담아서 JSON으로 보내는 것입니다.

따라서 해당 `@RequestBody`는 의도환 내용과 전혀 무관한 잘못된 내용입니다.
(Controller 에서 객체를 생성한 뒤, name을 설정하고 반환하면 됩니다.)

monorisk commented on src/main/java/com/minsuhas/springboot/controller/Mission5Controller.java:
```diff
@@ -0,0 +1,17 @@
+@RestController
+public class Mission5Controller {
+
+    @RequestMapping(value = "/resource/hello/{name}/json")
```
Spring Boot의 기본 설정에 따르면, 객체가 반환 되는 경우 제일 위에 언급된 jackson(클래스 이름은 ObjectMapper)를 이용해 JSON으로 변환 합니다.
(pom.xml에 추가한 의존성 쪽의 리뷰 참조)

위의 기본 설정 때문에, 이렇게 객체를 반환하는 경우는 `Content-Type`이  `application/json`으로 지정 되게 됩니다.

이것을 `produces`를 이용해 명시적으로 표현 해도 좋고, 묵시적으로 나타내도 됩니다.
(다만 하나의 통일성을 지향 하는 것이 좋습니다.)

monorisk commented on src/main/java/com/minsuhas/springboot/controller/User.java:
```diff
@@ -0,0 +1,13 @@
+package com.minsuhas.springboot.controller;
+
+public class User {
```
### 정보

이런식으로 데이터 컨테이너 외의 다른 역할을 하지 않는 것들을 Spring에서는 DTO 라고 합니다.
(Data Transfer Object)
(컨테이너: 이 맥락에서는 여러개의 변수를 하나의 객체에 담아 코드에서 다룰 수 있게 한다 정도로 생각하면 됩니다. 워낙 문맥에 따라 다양한 뜻을 가지는 단어라 해당 문맥을 파악하는게 중요합니다.)

DTO는 원시 타입의 데이터들을 감싸서 컨테이너로 만들기도 하며, 뒤에 언급될 Spring의 BO(Service), DAO(Repository) 계층간의 데이터를 주고 받는 객체로서 사용 됩니다.

지금은 이런식으로 행위가 없고 변수와 getter, setter만 가지고 있는게 DTO라는게 있다 정도만 알면 됩니다.
(물론 경우에 따라서는 getter or setter가 없을 수도 있습니다.)

monorisk commented on src/test/java/com/minsuhas/springboot/controller/Mission3ControllerTest.java:
```diff
@@ -0,0 +1,27 @@
+    public void testReturnName() throws Exception {
+        mockMvc.perform(get("/hello/minsu"))
+                .andExpect(status().isOk())
+                .andExpect(content().string("minsu"));
```
여기에 `Conent-Type`을 확인 해보는 코드가 있으면 좋을 것 같습니다.

monorisk replied:
### 추가적인 생각
* 이 테스트 코드가 왜 https://github.com/minsuhas/spring_boot/blob/master/src/main/java/com/minsuhas/springboot/controller/MinHasController.java#L17 의 Controller의 응답을 반환 하지 않는지 설명할 수 있나요?

monorisk commented on src/test/java/com/minsuhas/springboot/controller/Mission4ControllerTest.java:
```diff
@@ -0,0 +1,26 @@
+    @Test
+    public void testReturnImage() throws Exception {
+        mockMvc.perform(get("/resource/hello/image"))
+                .andExpect(status().isOk());
```
여기도 `Content-Type`을 확인하는 코드가 있으면 좋을 것 같습니다.

monorisk commented on src/test/java/com/minsuhas/springboot/controller/Mission5ControllerTest.java:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    public void testReturnJson() throws Exception {
+        String userJson = "{\"name\":\"minsu\"}";
+        mockMvc.perform(post("/resource/hello/minsu/json")
```
위에 리뷰된 내용에 맞춰, request body의 의미 없는 userJson은 삭제 해야 합니다.

monorisk commented on src/test/java/com/minsuhas/springboot/controller/Mission5ControllerTest.java:
```diff
@@ -0,0 +1,35 @@
+                .contentType(MediaType.APPLICATION_JSON_UTF8)
+                .accept(MediaType.APPLICATION_JSON_UTF8)
+                .content(userJson))
+            .andExpect(status().isOk())
```
여기도 `Content-Type`을 확인하는 코드가 있으면 좋을 것 같습니다.

monorisk commented:
PR(Pull Request)의 이름은 commit의 이름을 따라가는 경우가 있습니다. 

정식으로 리뷰의 대상이 되는 PR은 어떤 내용들이 추가 되었는지 제목으로 쉽게 유추할 수 있게 의미 있는 제목을 붙이는 것이 좋습니다.

PR이름을 WIPG -> 의미있는 이름 으로 변경해보아요

monorisk commented:
## 추가 미션

PR의 윗부분에 보면 `WIPG(working in progess)`라는 커밋이 들어 있는 것이 보이는데요, 사실 이런 의미없는 커밋 이름은 지양 해야 합니다.

PR을 열어서 merge를 요청하기 전에 커밋들을 잘 정리했어야 했는데 누락 되었습니다.

이번 PR은 커밋을 하나로 합쳐서 `mission 3 ~ 5 해결` 같은 이름을 짓는 것이 좋을 것 같은데요
어떻게 Git을 다뤄야 가능할까요?

* 힌트! `Squash and Merge`

monorisk commented:
> PR(Pull Request)의 이름은 commit의 이름을 따라가는 경우가 있습니다.
> 
> 정식으로 리뷰의 대상이 되는 PR은 어떤 내용들이 추가 되었는지 제목으로 쉽게 유추할 수 있게 의미 있는 제목을 붙이는 것이 좋습니다.

Min-Su committed:
#mission 3 ~ 5 해결
